test_conformance/subgroups/test_subgroup_ballot.cpp

```diff
@@ -0,0 +1,1089 @@
+    "    Type x = in[gid];\n"
+    "    uint4 mask = get_sub_group_lt_mask();"
+    "    out[gid] = mask;\n"
+    "}\n";
+
+static const char *get_subgroup_eq_mask_source =
+    "__kernel void test_get_sub_group_eq_mask(const __global Type *in, "
+    "__global int4 *xy, __global Type *out)\n"
+    "{\n"
+    "    int gid = get_global_id(0);\n"
+    "    XY(xy,gid);\n"
+    "    xy[gid].z = get_max_sub_group_size();\n"
+    "    Type x = in[gid];\n"
+    "    uint4 mask = get_sub_group_eq_mask();"
+    "    out[gid] = mask;\n"
+    "}\n";
+
+static const char *ballot_source =
+    "__kernel void test_sub_group_ballot(const __global Type *in, "
+    "__global int4 *xy, __global Type *out)\n"
+    "{\n"
+    "uint4 full_ballot = sub_group_ballot(1);\n"
+    "uint divergence_mask;\n"
+    "uint4 partial_ballot;\n"
+    "uint gid = get_global_id(0);"
+    "XY(xy,gid);\n"
+    "if (get_sub_group_local_id() & 1) {\n"
+    "    divergence_mask = 0xaaaaaaaa;\n"
+    "    partial_ballot = sub_group_ballot(1);\n"
+    "} else {\n"
+    "    divergence_mask = 0x55555555;\n"
+    "    partial_ballot = sub_group_ballot(1);\n"
+    "}\n"
+    " size_t lws = get_local_size(0);\n"
+    "uint4 masked_ballot = full_ballot;\n"
+    "masked_ballot.x &= divergence_mask;\n"
+    "masked_ballot.y &= divergence_mask;\n"
+    "masked_ballot.z &= divergence_mask;\n"
+    "masked_ballot.w &= divergence_mask;\n"
+    "out[gid] = all(masked_ballot == partial_ballot);\n"
+
+    "} \n";
+
+static const char *ballot_source_inverse =
+    "__kernel void test_sub_group_ballot_inverse(const __global "
+    "Type *in, "
+    "__global int4 *xy, __global Type *out)\n"
+    "{\n"
+    "    int gid = get_global_id(0);\n"
+    "    XY(xy,gid);\n"
+    "    Type x = in[gid];\n"
+    "    uint4 value = (uint4)(10,0,0,0);\n"
+    "    if (get_sub_group_local_id() & 1) {"
+    "        uint4 partial_ballot_mask = "
+    "(uint4)(0xAAAAAAAA,0xAAAAAAAA,0xAAAAAAAA,0xAAAAAAAA);"
+    "        if (sub_group_inverse_ballot(partial_ballot_mask)) {\n"
+    "            value = (uint4)(1,0,0,1);\n"
+    "        } else {\n"
+    "            value = (uint4)(0,0,0,1);\n"
+    "        }\n"
+    "    } else {\n"
+    "       uint4 partial_ballot_mask = "
+    "(uint4)(0x55555555,0x55555555,0x55555555,0x55555555);"
+    "        if (sub_group_inverse_ballot(partial_ballot_mask)) {\n"
+    "            value = (uint4)(1,0,0,2);\n"
+    "        } else {\n"
+    "            value = (uint4)(0,0,0,2);\n"
+    "        }\n"
+    "    }\n"
+    "    out[gid] = value;\n"
+    "}\n";
+
+static const char *ballot_bit_extract_source =
+    "__kernel void test_sub_group_ballot_bit_extract(const __global Type *in, "
+    "__global int4 *xy, __global Type *out)\n"
+    "{\n"
+    "    int gid = get_global_id(0);\n"
+    "    XY(xy,gid);\n"
+    "    Type x = in[gid];\n"
+    "    uint index = xy[gid].z;\n"
+    "    uint4 value = (uint4)(10,0,0,0);\n"
+    "    if (get_sub_group_local_id() & 1) {"
```

Comment:
As per sub_group_ballot_inverse(), sub_group_ballot_bit_extract() is not sensitive to uniformity of control flow. Why is divergence tested here?

Reply:
Will be solved by https://github.com/KhronosGroup/OpenCL-CTS/issues/1207

Reply:
Agreed.